### PR TITLE
[IMP] mail: mark first welcome message as read for the admin

### PR DESCRIPTION
**Pull request by dhp-odoo**
**PURPOSE**

Actually, We don't want to distract the admin with an unnecessary notification
on General Channel. The main focus should be on odoobot's tour, not on the
general channel's welcome message. And it is also not a good idea to remove the
general channel as it would reduce the discoverability of the feature.

**SPECIFICATION**

So, we are marking the first message as read at least for the admin user.
therefore admin can not distract over the unnecessary welcome notifications and
focus more on the rest of the functionality.

PS: we can not use "_set_last_seen_message" method as it takes partner_id of SUPERUSER 
because this method calls once during installation and installation is taken place by SUPERUSER
But we required admin users to mark the first message as read.

**Task : 2442023**

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
Maybe simple remove unnecessary messages, allowing to clean code and data. Init hook should not be used for that kind of demo-data issue. Moreover it will not work on saas as post init hooks are not called.

**Comment by seb-odoo** [comment]
Yeah this is one of those onboarding tasks...
3 months ago we were asked to reintroduce these messages (that were otherwise bugged and not showing) because the channel "looked empty and useless" without them.

**Comment by seb-odoo** [comment]
The main problem is that we use this channel to post technical messages so we can't remove it from the db, otherwise the initial suggestion was just to remove it.

File: addons/mail/models/mail_channel.py [hunk 1]
```diff
@@ -1133,3 +1133,21 @@ def _execute_command_who(self, **kwargs):
...
+            if channel and message:
+                channel_partner = self.env['mail.channel.partner'].search([
+                    ('channel_id', '=', channel.id),
+                    ('partner_id', '=', user.partner_id.id),
+                    ('seen_message_id', '=', False),
+                ])
+                channel_partner.write({
+                    'fetched_message_id': message.id,
+                    'seen_message_id': message.id,
+                })
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1153]: You can mark it as read for all existing users at that time, not just the admin: so remove `partner_id` and `seen_message_id` from the domain and remove the loop, and just write on all channel_partners of this channel in one query.

Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:1153]: Also we probably want to fetch the last message from the channel dynamically, instead of relying on `ref`.

Comment (dhp-odoo) [reply addons/mail/models/mail_channel.py:1153]: @sea-odoo . you mean we have to mark as read whole "general" channel? as there are total 5 messages in the channel.
And as per the task need to mark as read first message of the "general" channel.

Comment (sea-odoo) [reply addons/mail/models/mail_channel.py:1153]: @seb-odoo  Wrong tag, this one is for you 
f.y.i : @dhp-odoo

Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:1153]: The goal is to remove the notification and the counter after a fresh install, so all messages present in the channel have to be marked as read for all users.

Comment (dhp-odoo) [reply addons/mail/models/mail_channel.py]: @seb-odoo just have a look.

File: addons/mail/models/mail_channel.py [hunk 2]
```diff
@@ -1133,3 +1133,13 @@ def _execute_command_who(self, **kwargs):
             msg = _("Users in this channel: %(members)s %(dots)s and you.", members=", ".join(members), dots=dots)
 
         self._send_transient_message(partner, msg)
+
+    @api.model
+    def _mark_seen_general_channel(self):
+        channel = self.env.ref('mail.channel_all_employees', raise_if_not_found=True)
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1139]: `raise_if_not_found` `True` is the default value, no need to specify it.

File: addons/mail/models/mail_channel.py [hunk 3]
```diff
@@ -1133,3 +1133,13 @@ def _execute_command_who(self, **kwargs):
...
+        last_message_id = channel._channel_last_message_ids()
+        if channel and last_message_id:
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1141]: no need to make `if channel` if you raise if it's not found

File: addons/mail/models/mail_channel.py [hunk 4]
```diff
@@ -1133,3 +1133,13 @@ def _execute_command_who(self, **kwargs):
             msg = _("Users in this channel: %(members)s %(dots)s and you.", members=", ".join(members), dots=dots)
 
         self._send_transient_message(partner, msg)
+
+    @api.model
+    def _mark_seen_general_channel(self):
+        channel = self.env.ref('mail.channel_all_employees', raise_if_not_found=True)
+        last_message_id = channel._channel_last_message_ids()
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1140]: take the target value early to avoid manipulating the list of dict everywhere, and adapt the rest accordingly
```suggestion
        last_message_data = channel._channel_last_message_ids()[0]
        last_message_id = last_message_data.get("message_id") if last_message_data else False
```

File: addons/mail/models/mail_channel.py [hunk 5]
```diff
@@ -1133,3 +1133,13 @@ def _execute_command_who(self, **kwargs):
             msg = _("Users in this channel: %(members)s %(dots)s and you.", members=", ".join(members), dots=dots)
 
         self._send_transient_message(partner, msg)
+
+    @api.model
+    def _mark_seen_general_channel(self):
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1138]: I would use the technical name of the channel here, and specify it's a mark as seen for everyone because that is very unusual and specific.
```suggestion
    def _mark_channel_all_employees_as_seen_for_everyone(self):
```

File: addons/mail/models/mail_channel.py [hunk 6]
```diff
@@ -1133,3 +1133,13 @@ def _execute_command_who(self, **kwargs):
...
+        channel = self.env.ref('mail.channel_all_employees', raise_if_not_found=True)
+        last_message_id = channel._channel_last_message_ids()
+        if channel and last_message_id:
+            channel.channel_last_seen_partner_ids.write({
+                'fetched_message_id': last_message_id[0] and last_message_id[0].get("message_id"),
+                'seen_message_id': last_message_id[0] and last_message_id[0].get("message_id"),
+            })
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:1145]: no need to guard `last_message_id` on each line, just the top level `if` is enough

**Review by seb-odoo — COMMENTED** [review]
Thank you, I think it should do what we need now. Some comments for code cleanup.

**Comment by tde-banana-odoo** [comment]
:vomiting_face:

**Comment by seb-odoo** [comment]
Also please adapt commit/PR description accordingly.

**Comment by tde-banana-odoo** [comment]
"perfect onboarding yay" 8-) I know those tasks ... not always easy to cancel them :D

**Comment by seb-odoo** [comment]
> "perfect onboarding yay" 8-) I know those tasks ... not always easy to cancel them :D

I don't feel it's bad enough to be worth fighting against it, but feel free if you do.

**Comment by tde-banana-odoo** [comment]
Well I think this whole PR makes no sense. I think this is important to battle because it adds unnecessary code to maintain and test. That is a technical debt.

Demo mode: you have demo messages, that's the whole purpose of it. If you don't want people to see them, don't log them.

Data mode: you should not have demo messages. If you have data messages, probably they makes sense. If not they should not be logged.

**Comment by tde-banana-odoo** [comment]
Another way round would be to manually create a channel.partner entry for admin and set it last fetched message to "some seconds in the future" as a classic xml demo record. But please don't add python code to handle demo data at install.

**Comment by seb-odoo** [comment]
The target is not demo data though, it is real data. We don't introduce code for demo, we introduce code to actually have any (real) DB set up how we would want it.

There is a welcome message in real data for this channel, it's just a welcome message to explain the purpose of the channel (and to avoid having it empty which is would understandably be worse for on boarding).

The argument is that it's useful to see the message in the channel, but it's not useful to have a notification for it.

File: addons/mail/data/mail_channel_demo.xml [hunk 7]
```diff
@@ -163,5 +163,7 @@
             <field name="date" eval="(DateTime.today() - timedelta(days=3)).strftime('%Y-%m-%d %H:%M')"/>
         </record>
 
+        <function model="mail.channel" name="_mark_seen_general_channel"/>
```
Comment (seb-odoo) [inline addons/mail/data/mail_channel_demo.xml:166]: According to what I just posted above, I didn't pay attention but this call should actually be in real data, not demo data.

**Comment by tde-banana-odoo** [comment]
Then consider creating manually a channel partner entry and update it in data. Don't add python code for that. Especially that this message could be replaced by a group description displayed in Discuss instead of a dummy message that nobody will read once several messages have been posted.

**Comment by seb-odoo** [comment]
Fine for me if we can do it with a manual channel partner entry, but auto-subscribe from groups channel might get in the way. Worth to investigate.

**Comment by seb-odoo** [comment]
@dhp-odoo could you please try it this way? In `mail_channel_data.xml` you would create manually the `mail.channel.partner` entry for "admin/general channel" after the message has been created, with the fetch and seen values according to id of created message.
This will probably require setting `<field name="group_ids" eval="[(4, ref('base.group_user'))]"/>` on the channel after all of this is done.

**Comment by dhp-odoo** [comment]
> @dhp-odoo could you please try it this way? In `mail_channel_data.xml` you would create manually the `mail.channel.partner` entry for "admin/general channel" after the message has been created, with the fetch and seen values according to id of created message.
> This will probably require setting `<field name="group_ids" eval="[(4, ref('base.group_user'))]"/>` on the channel after all of this is done.

I can understand what you want to implement, but there are some clarifications required as below:

1) As you told, create the manual entry for the `mail.channel.partner` in the `mail_channel_data.xml` file. but the thing is mail.channel.partner is creating while the channel is creating (refer `_subscribe_users` method in mail_channel.py). and there is no constrain for the duplicate entry. Now if we create the manual entry from the XML data. then it will create a separate duplicate entry for the same partner and channel. so should I consider updating the newly created `mail.channel.partner` entry?

2) Now in the `mail_channel_data.xml` file, there is only one message "welcome message" but as per the spec describe above, we have to mark all the messages as read of the same channel. and some more demo messages are created in the "mail_channel_demo.xml". now I can not get the last possible message of the channel in this file where you suggested.

3) `This will probably require setting `<field name="group_ids" eval="[(4, ref('base.group_user'))]"/>` on the channel after all of this is done.`  >> I didn't get your point.

**Comment by seb-odoo** [comment]
For 1. the answer is 4. the `_subscribe_users` is not called if the channel does not have value for the `group_ids` field. So the way to avoid it is by removing the field from the data at the top (so that it is not set during create), and only add the field after the channel.partner has been created. (this way when trying it create from `_subscribe_users` it should find it already exists).

2. You can ignore messages from demo now, we only target onboarding on a new empty DB. If we wanted to target demo too, we would do the same trick in the demo data file, where we update the channel partner to the last message id of the last demo message. But you don't have to do this now.

3. It's fine to ignore other users too, only admin exists when creating a non-demo DB.

In summary we are just reverting to a simpler solution where we only target admin and only for the welcome message, with the goal being no python code and only data change.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
